From 893e0a4b859a1932d729c5abb6d67b8b4410af4a Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Mon, 7 Sep 2026 13:19:42 +0000
Subject: [PATCH] Chainguard Cryptographic Hardening

Unregister weak, inscure, deprecated, broken algorithms from the
default provider. All of these are unused, not part of the FIPS
provider, deprecated by IETF and IANA.

This is done at the provider level, to otherwise keep stable and
compatible ABI and API.
---
 crypto/ec/ec_curve.c            | 179 --------------------------------
 providers/common/capabilities.c |  19 ----
 providers/defltprov.c           | 157 ----------------------------
 ssl/ssl_ciph.c                  |  18 ++++
 ssl/t1_lib.c                    | 104 +++++++++++++++----
 5 files changed, 101 insertions(+), 376 deletions(-)

diff --git a/crypto/ec/ec_curve.c b/crypto/ec/ec_curve.c
index e63d89bae7..62ef08f460 100644
--- a/crypto/ec/ec_curve.c
+++ b/crypto/ec/ec_curve.c
@@ -2615,32 +2615,6 @@ static const ec_list_element curve_list[] = {
 static const ec_list_element curve_list[] = {
     /* prime field curves */
     /* secg curves */
-    { NID_secp112r1, &_EC_SECG_PRIME_112R1.h, 0,
-        "SECG/WTLS curve over a 112 bit prime field" },
-    { NID_secp112r2, &_EC_SECG_PRIME_112R2.h, 0,
-        "SECG curve over a 112 bit prime field" },
-    { NID_secp128r1, &_EC_SECG_PRIME_128R1.h, 0,
-        "SECG curve over a 128 bit prime field" },
-    { NID_secp128r2, &_EC_SECG_PRIME_128R2.h, 0,
-        "SECG curve over a 128 bit prime field" },
-    { NID_secp160k1, &_EC_SECG_PRIME_160K1.h, 0,
-        "SECG curve over a 160 bit prime field" },
-    { NID_secp160r1, &_EC_SECG_PRIME_160R1.h, 0,
-        "SECG curve over a 160 bit prime field" },
-    { NID_secp160r2, &_EC_SECG_PRIME_160R2.h, 0,
-        "SECG/WTLS curve over a 160 bit prime field" },
-    /* SECG secp192r1 is the same as X9.62 prime192v1 and hence omitted */
-    { NID_secp192k1, &_EC_SECG_PRIME_192K1.h, 0,
-        "SECG curve over a 192 bit prime field" },
-    { NID_secp224k1, &_EC_SECG_PRIME_224K1.h, 0,
-        "SECG curve over a 224 bit prime field" },
-#ifndef OPENSSL_NO_EC_NISTP_64_GCC_128
-    { NID_secp224r1, &_EC_NIST_PRIME_224.h, EC_GFp_nistp224_method,
-        "NIST/SECG curve over a 224 bit prime field" },
-#else
-    { NID_secp224r1, &_EC_NIST_PRIME_224.h, 0,
-        "NIST/SECG curve over a 224 bit prime field" },
-#endif
     { NID_secp256k1, &_EC_SECG_PRIME_256K1.h, 0,
         "SECG curve over a 256 bit prime field" },
     /* SECG secp256r1 is the same as X9.62 prime256v1 and hence omitted */
@@ -2663,18 +2637,6 @@ static const ec_list_element curve_list[] = {
 #endif
         "NIST/SECG curve over a 521 bit prime field" },
     /* X9.62 curves */
-    { NID_X9_62_prime192v1, &_EC_NIST_PRIME_192.h, 0,
-        "NIST/X9.62/SECG curve over a 192 bit prime field" },
-    { NID_X9_62_prime192v2, &_EC_X9_62_PRIME_192V2.h, 0,
-        "X9.62 curve over a 192 bit prime field" },
-    { NID_X9_62_prime192v3, &_EC_X9_62_PRIME_192V3.h, 0,
-        "X9.62 curve over a 192 bit prime field" },
-    { NID_X9_62_prime239v1, &_EC_X9_62_PRIME_239V1.h, 0,
-        "X9.62 curve over a 239 bit prime field" },
-    { NID_X9_62_prime239v2, &_EC_X9_62_PRIME_239V2.h, 0,
-        "X9.62 curve over a 239 bit prime field" },
-    { NID_X9_62_prime239v3, &_EC_X9_62_PRIME_239V3.h, 0,
-        "X9.62 curve over a 239 bit prime field" },
     { NID_X9_62_prime256v1, &_EC_X9_62_PRIME_256V1.h,
 #if defined(ECP_NISTZ256_ASM)
         EC_GFp_nistz256_method,
@@ -2686,154 +2648,13 @@ static const ec_list_element curve_list[] = {
         0,
 #endif
         "X9.62/SECG curve over a 256 bit prime field" },
-#ifndef OPENSSL_NO_EC2M
-    /* characteristic two field curves */
-    /* NIST/SECG curves */
-    { NID_sect113r1, &_EC_SECG_CHAR2_113R1.h, 0,
-        "SECG curve over a 113 bit binary field" },
-    { NID_sect113r2, &_EC_SECG_CHAR2_113R2.h, 0,
-        "SECG curve over a 113 bit binary field" },
-    { NID_sect131r1, &_EC_SECG_CHAR2_131R1.h, 0,
-        "SECG/WTLS curve over a 131 bit binary field" },
-    { NID_sect131r2, &_EC_SECG_CHAR2_131R2.h, 0,
-        "SECG curve over a 131 bit binary field" },
-    { NID_sect163k1, &_EC_NIST_CHAR2_163K.h, 0,
-        "NIST/SECG/WTLS curve over a 163 bit binary field" },
-    { NID_sect163r1, &_EC_SECG_CHAR2_163R1.h, 0,
-        "SECG curve over a 163 bit binary field" },
-    { NID_sect163r2, &_EC_NIST_CHAR2_163B.h, 0,
-        "NIST/SECG curve over a 163 bit binary field" },
-    { NID_sect193r1, &_EC_SECG_CHAR2_193R1.h, 0,
-        "SECG curve over a 193 bit binary field" },
-    { NID_sect193r2, &_EC_SECG_CHAR2_193R2.h, 0,
-        "SECG curve over a 193 bit binary field" },
-    { NID_sect233k1, &_EC_NIST_CHAR2_233K.h, 0,
-        "NIST/SECG/WTLS curve over a 233 bit binary field" },
-    { NID_sect233r1, &_EC_NIST_CHAR2_233B.h, 0,
-        "NIST/SECG/WTLS curve over a 233 bit binary field" },
-    { NID_sect239k1, &_EC_SECG_CHAR2_239K1.h, 0,
-        "SECG curve over a 239 bit binary field" },
-    { NID_sect283k1, &_EC_NIST_CHAR2_283K.h, 0,
-        "NIST/SECG curve over a 283 bit binary field" },
-    { NID_sect283r1, &_EC_NIST_CHAR2_283B.h, 0,
-        "NIST/SECG curve over a 283 bit binary field" },
-    { NID_sect409k1, &_EC_NIST_CHAR2_409K.h, 0,
-        "NIST/SECG curve over a 409 bit binary field" },
-    { NID_sect409r1, &_EC_NIST_CHAR2_409B.h, 0,
-        "NIST/SECG curve over a 409 bit binary field" },
-    { NID_sect571k1, &_EC_NIST_CHAR2_571K.h, 0,
-        "NIST/SECG curve over a 571 bit binary field" },
-    { NID_sect571r1, &_EC_NIST_CHAR2_571B.h, 0,
-        "NIST/SECG curve over a 571 bit binary field" },
-    /* X9.62 curves */
-    { NID_X9_62_c2pnb163v1, &_EC_X9_62_CHAR2_163V1.h, 0,
-        "X9.62 curve over a 163 bit binary field" },
-    { NID_X9_62_c2pnb163v2, &_EC_X9_62_CHAR2_163V2.h, 0,
-        "X9.62 curve over a 163 bit binary field" },
-    { NID_X9_62_c2pnb163v3, &_EC_X9_62_CHAR2_163V3.h, 0,
-        "X9.62 curve over a 163 bit binary field" },
-    { NID_X9_62_c2pnb176v1, &_EC_X9_62_CHAR2_176V1.h, 0,
-        "X9.62 curve over a 176 bit binary field" },
-    { NID_X9_62_c2tnb191v1, &_EC_X9_62_CHAR2_191V1.h, 0,
-        "X9.62 curve over a 191 bit binary field" },
-    { NID_X9_62_c2tnb191v2, &_EC_X9_62_CHAR2_191V2.h, 0,
-        "X9.62 curve over a 191 bit binary field" },
-    { NID_X9_62_c2tnb191v3, &_EC_X9_62_CHAR2_191V3.h, 0,
-        "X9.62 curve over a 191 bit binary field" },
-    { NID_X9_62_c2pnb208w1, &_EC_X9_62_CHAR2_208W1.h, 0,
-        "X9.62 curve over a 208 bit binary field" },
-    { NID_X9_62_c2tnb239v1, &_EC_X9_62_CHAR2_239V1.h, 0,
-        "X9.62 curve over a 239 bit binary field" },
-    { NID_X9_62_c2tnb239v2, &_EC_X9_62_CHAR2_239V2.h, 0,
-        "X9.62 curve over a 239 bit binary field" },
-    { NID_X9_62_c2tnb239v3, &_EC_X9_62_CHAR2_239V3.h, 0,
-        "X9.62 curve over a 239 bit binary field" },
-    { NID_X9_62_c2pnb272w1, &_EC_X9_62_CHAR2_272W1.h, 0,
-        "X9.62 curve over a 272 bit binary field" },
-    { NID_X9_62_c2pnb304w1, &_EC_X9_62_CHAR2_304W1.h, 0,
-        "X9.62 curve over a 304 bit binary field" },
-    { NID_X9_62_c2tnb359v1, &_EC_X9_62_CHAR2_359V1.h, 0,
-        "X9.62 curve over a 359 bit binary field" },
-    { NID_X9_62_c2pnb368w1, &_EC_X9_62_CHAR2_368W1.h, 0,
-        "X9.62 curve over a 368 bit binary field" },
-    { NID_X9_62_c2tnb431r1, &_EC_X9_62_CHAR2_431R1.h, 0,
-        "X9.62 curve over a 431 bit binary field" },
-    /*
-     * the WAP/WTLS curves [unlike SECG, spec has its own OIDs for curves
-     * from X9.62]
-     */
-    { NID_wap_wsg_idm_ecid_wtls1, &_EC_WTLS_1.h, 0,
-        "WTLS curve over a 113 bit binary field" },
-    { NID_wap_wsg_idm_ecid_wtls3, &_EC_NIST_CHAR2_163K.h, 0,
-        "NIST/SECG/WTLS curve over a 163 bit binary field" },
-    { NID_wap_wsg_idm_ecid_wtls4, &_EC_SECG_CHAR2_113R1.h, 0,
-        "SECG curve over a 113 bit binary field" },
-    { NID_wap_wsg_idm_ecid_wtls5, &_EC_X9_62_CHAR2_163V1.h, 0,
-        "X9.62 curve over a 163 bit binary field" },
-#endif
-    { NID_wap_wsg_idm_ecid_wtls6, &_EC_SECG_PRIME_112R1.h, 0,
-        "SECG/WTLS curve over a 112 bit prime field" },
-    { NID_wap_wsg_idm_ecid_wtls7, &_EC_SECG_PRIME_160R2.h, 0,
-        "SECG/WTLS curve over a 160 bit prime field" },
-    { NID_wap_wsg_idm_ecid_wtls8, &_EC_WTLS_8.h, 0,
-        "WTLS curve over a 112 bit prime field" },
-    { NID_wap_wsg_idm_ecid_wtls9, &_EC_WTLS_9.h, 0,
-        "WTLS curve over a 160 bit prime field" },
-#ifndef OPENSSL_NO_EC2M
-    { NID_wap_wsg_idm_ecid_wtls10, &_EC_NIST_CHAR2_233K.h, 0,
-        "NIST/SECG/WTLS curve over a 233 bit binary field" },
-    { NID_wap_wsg_idm_ecid_wtls11, &_EC_NIST_CHAR2_233B.h, 0,
-        "NIST/SECG/WTLS curve over a 233 bit binary field" },
-#endif
-    { NID_wap_wsg_idm_ecid_wtls12, &_EC_WTLS_12.h, 0,
-        "WTLS curve over a 224 bit prime field" },
-#ifndef OPENSSL_NO_EC2M
-    /* IPSec curves */
-    { NID_ipsec3, &_EC_IPSEC_155_ID3.h, 0,
-        "\n\tIPSec/IKE/Oakley curve #3 over a 155 bit binary field.\n"
-        "\tNot suitable for ECDSA.\n\tQuestionable extension field!" },
-    { NID_ipsec4, &_EC_IPSEC_185_ID4.h, 0,
-        "\n\tIPSec/IKE/Oakley curve #4 over a 185 bit binary field.\n"
-        "\tNot suitable for ECDSA.\n\tQuestionable extension field!" },
-#endif
     /* brainpool curves */
-    { NID_brainpoolP160r1, &_EC_brainpoolP160r1.h, 0,
-        "RFC 5639 curve over a 160 bit prime field" },
-    { NID_brainpoolP160t1, &_EC_brainpoolP160t1.h, 0,
-        "RFC 5639 curve over a 160 bit prime field" },
-    { NID_brainpoolP192r1, &_EC_brainpoolP192r1.h, 0,
-        "RFC 5639 curve over a 192 bit prime field" },
-    { NID_brainpoolP192t1, &_EC_brainpoolP192t1.h, 0,
-        "RFC 5639 curve over a 192 bit prime field" },
-    { NID_brainpoolP224r1, &_EC_brainpoolP224r1.h, 0,
-        "RFC 5639 curve over a 224 bit prime field" },
-    { NID_brainpoolP224t1, &_EC_brainpoolP224t1.h, 0,
-        "RFC 5639 curve over a 224 bit prime field" },
     { NID_brainpoolP256r1, &_EC_brainpoolP256r1.h, 0,
         "RFC 5639 curve over a 256 bit prime field" },
-    { NID_brainpoolP256t1, &_EC_brainpoolP256t1.h, 0,
-        "RFC 5639 curve over a 256 bit prime field" },
-    { NID_brainpoolP320r1, &_EC_brainpoolP320r1.h, 0,
-        "RFC 5639 curve over a 320 bit prime field" },
-    { NID_brainpoolP320t1, &_EC_brainpoolP320t1.h, 0,
-        "RFC 5639 curve over a 320 bit prime field" },
     { NID_brainpoolP384r1, &_EC_brainpoolP384r1.h, 0,
         "RFC 5639 curve over a 384 bit prime field" },
-    { NID_brainpoolP384t1, &_EC_brainpoolP384t1.h, 0,
-        "RFC 5639 curve over a 384 bit prime field" },
     { NID_brainpoolP512r1, &_EC_brainpoolP512r1.h, 0,
         "RFC 5639 curve over a 512 bit prime field" },
-    { NID_brainpoolP512t1, &_EC_brainpoolP512t1.h, 0,
-        "RFC 5639 curve over a 512 bit prime field" },
-#ifndef OPENSSL_NO_SM2
-    { NID_sm2, &_EC_sm2p256v1.h,
-#ifdef ECP_SM2P256_ASM
-        EC_GFp_sm2p256_method,
-#else
-        0,
-#endif
-        "SM2 curve over a 256 bit prime field" },
-#endif
 };
 #endif /* FIPS_MODULE */
 
diff --git a/providers/common/capabilities.c b/providers/common/capabilities.c
index b97d4d6bcd..b34ec10b85 100644
--- a/providers/common/capabilities.c
+++ b/providers/common/capabilities.c
@@ -171,11 +171,6 @@ static const OSSL_PARAM param_group_list[][11] = {
     TLS_GROUP_ENTRY("secp521r1", "secp521r1", "EC", 24),
     TLS_GROUP_ENTRY("P-521", "secp521r1", "EC", 24), /* Alias of above */
 #endif /* OPENSSL_NO_EC */
-#ifndef OPENSSL_NO_DH
-    /* Security bit values for FFDHE groups are as per RFC 7919 */
-    TLS_GROUP_ENTRY("ffdhe2048", "ffdhe2048", "DH", 33),
-    TLS_GROUP_ENTRY("ffdhe3072", "ffdhe3072", "DH", 34),
-#endif
 #if !defined(OPENSSL_NO_ML_KEM)
     /* https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-8 */
     TLS_GROUP_ENTRY("MLKEM512", "", "ML-KEM-512", 38),
@@ -190,23 +185,12 @@ static const OSSL_PARAM param_group_list[][11] = {
     TLS_GROUP_ENTRY("brainpoolP256r1tls13", "brainpoolP256r1", "EC", 30),
     TLS_GROUP_ENTRY("brainpoolP384r1tls13", "brainpoolP384r1", "EC", 31),
     TLS_GROUP_ENTRY("brainpoolP512r1tls13", "brainpoolP512r1", "EC", 32),
-#ifndef OPENSSL_NO_SM2
-    TLS_GROUP_ENTRY("curveSM2", "SM2", "curveSM2", 44),
-#if !defined(OPENSSL_NO_ML_KEM)
-    TLS_GROUP_ENTRY("curveSM2MLKEM768", "", "curveSM2MLKEM768", 45),
-#endif
-#endif
 #endif
 #ifndef OPENSSL_NO_ML_KEM
     TLS_GROUP_ENTRY("SecP256r1MLKEM768", "", "SecP256r1MLKEM768", 42),
     TLS_GROUP_ENTRY("SecP384r1MLKEM1024", "", "SecP384r1MLKEM1024", 43),
 #endif
 #endif
-#ifndef OPENSSL_NO_DH
-    TLS_GROUP_ENTRY("ffdhe4096", "ffdhe4096", "DH", 35),
-    TLS_GROUP_ENTRY("ffdhe6144", "ffdhe6144", "DH", 36),
-    TLS_GROUP_ENTRY("ffdhe8192", "ffdhe8192", "DH", 37),
-#endif
 #ifndef OPENSSL_NO_TLS_DEPRECATED_EC
 #ifndef OPENSSL_NO_EC2M
     TLS_GROUP_ENTRY("sect163k1", "sect163k1", "EC", 0),
@@ -329,9 +313,6 @@ static const OSSL_PARAM param_sigalg_list[][10] = {
     TLS_SIGALG_ENTRY("mldsa65", "ML-DSA-65", "2.16.840.1.101.3.4.3.18", 1),
     TLS_SIGALG_ENTRY("mldsa87", "ML-DSA-87", "2.16.840.1.101.3.4.3.19", 2),
 #endif
-#if !defined(FIPS_MODULE) && !defined(OPENSSL_NO_SM2) && !defined(OPENSSL_NO_SM3)
-    TLS_SIGALG_ENTRY("sm2sig_sm3", "SM2", "1.2.156.10197.1 501", 3),
-#endif
 };
 #endif
 
diff --git a/providers/defltprov.c b/providers/defltprov.c
index db9f84748a..79e9776750 100644
--- a/providers/defltprov.c
+++ b/providers/defltprov.c
@@ -97,21 +97,17 @@ static int deflt_get_params(void *provctx, OSSL_PARAM params[])
 static const OSSL_ALGORITHM deflt_digests[] = {
     /* Our primary name:NIST name[:our older names] */
     { PROV_NAMES_SHA1, "provider=default", ossl_sha1_functions },
-    { PROV_NAMES_SHA2_224, "provider=default", ossl_sha224_functions },
     { PROV_NAMES_SHA2_256, "provider=default", ossl_sha256_functions },
     { PROV_NAMES_SHA2_256_192, "provider=default", ossl_sha256_192_internal_functions },
     { PROV_NAMES_SHA2_384, "provider=default", ossl_sha384_functions },
     { PROV_NAMES_SHA2_512, "provider=default", ossl_sha512_functions },
-    { PROV_NAMES_SHA2_512_224, "provider=default", ossl_sha512_224_functions },
     { PROV_NAMES_SHA2_512_256, "provider=default", ossl_sha512_256_functions },
 
     /* We agree with NIST here, so one name only */
-    { PROV_NAMES_SHA3_224, "provider=default", ossl_sha3_224_functions },
     { PROV_NAMES_SHA3_256, "provider=default", ossl_sha3_256_functions },
     { PROV_NAMES_SHA3_384, "provider=default", ossl_sha3_384_functions },
     { PROV_NAMES_SHA3_512, "provider=default", ossl_sha3_512_functions },
 
-    { PROV_NAMES_KECCAK_224, "provider=default", ossl_keccak_224_functions },
     { PROV_NAMES_KECCAK_256, "provider=default", ossl_keccak_256_functions },
     { PROV_NAMES_KECCAK_384, "provider=default", ossl_keccak_384_functions },
     { PROV_NAMES_KECCAK_512, "provider=default", ossl_keccak_512_functions },
@@ -144,19 +140,10 @@ static const OSSL_ALGORITHM deflt_digests[] = {
     { PROV_NAMES_BLAKE2B_512, "provider=default", ossl_blake2b512_functions },
 #endif /* OPENSSL_NO_BLAKE2 */
 
-#ifndef OPENSSL_NO_SM3
-    { PROV_NAMES_SM3, "provider=default", ossl_sm3_functions },
-#endif /* OPENSSL_NO_SM3 */
-
 #ifndef OPENSSL_NO_MD5
     { PROV_NAMES_MD5, "provider=default", ossl_md5_functions },
-    { PROV_NAMES_MD5_SHA1, "provider=default", ossl_md5_sha1_functions },
 #endif /* OPENSSL_NO_MD5 */
 
-#ifndef OPENSSL_NO_RMD160
-    { PROV_NAMES_RIPEMD_160, "provider=default", ossl_ripemd160_functions },
-#endif /* OPENSSL_NO_RMD160 */
-
     { PROV_NAMES_NULL, "provider=default", ossl_nullmd_functions },
 #ifndef OPENSSL_NO_ML_DSA
     { PROV_NAMES_ML_DSA_MU, "provider=default", ossl_ml_dsa_mu_functions },
@@ -192,11 +179,6 @@ static const OSSL_ALGORITHM_CAPABLE deflt_ciphers[] = {
     ALG(PROV_NAMES_AES_128_CTR, ossl_aes128ctr_functions),
     ALG(PROV_NAMES_AES_256_XTS, ossl_aes256xts_functions),
     ALG(PROV_NAMES_AES_128_XTS, ossl_aes128xts_functions),
-#ifndef OPENSSL_NO_OCB
-    ALG(PROV_NAMES_AES_256_OCB, ossl_aes256ocb_functions),
-    ALG(PROV_NAMES_AES_192_OCB, ossl_aes192ocb_functions),
-    ALG(PROV_NAMES_AES_128_OCB, ossl_aes128ocb_functions),
-#endif /* OPENSSL_NO_OCB */
 #ifndef OPENSSL_NO_SIV
     ALG(PROV_NAMES_AES_128_SIV, ossl_aes128siv_functions),
     ALG(PROV_NAMES_AES_192_SIV, ossl_aes192siv_functions),
@@ -249,84 +231,6 @@ static const OSSL_ALGORITHM_CAPABLE deflt_ciphers[] = {
         ossl_cipher_capable_aes_cbc_hmac_sha512_etm),
     ALGC(PROV_NAMES_AES_256_CBC_HMAC_SHA512_ETM, ossl_aes256cbc_hmac_sha512_etm_functions,
         ossl_cipher_capable_aes_cbc_hmac_sha512_etm),
-#ifndef OPENSSL_NO_ARIA
-    ALG(PROV_NAMES_ARIA_256_GCM, ossl_aria256gcm_functions),
-    ALG(PROV_NAMES_ARIA_192_GCM, ossl_aria192gcm_functions),
-    ALG(PROV_NAMES_ARIA_128_GCM, ossl_aria128gcm_functions),
-    ALG(PROV_NAMES_ARIA_256_CCM, ossl_aria256ccm_functions),
-    ALG(PROV_NAMES_ARIA_192_CCM, ossl_aria192ccm_functions),
-    ALG(PROV_NAMES_ARIA_128_CCM, ossl_aria128ccm_functions),
-    ALG(PROV_NAMES_ARIA_256_ECB, ossl_aria256ecb_functions),
-    ALG(PROV_NAMES_ARIA_192_ECB, ossl_aria192ecb_functions),
-    ALG(PROV_NAMES_ARIA_128_ECB, ossl_aria128ecb_functions),
-    ALG(PROV_NAMES_ARIA_256_CBC, ossl_aria256cbc_functions),
-    ALG(PROV_NAMES_ARIA_192_CBC, ossl_aria192cbc_functions),
-    ALG(PROV_NAMES_ARIA_128_CBC, ossl_aria128cbc_functions),
-    ALG(PROV_NAMES_ARIA_256_OFB, ossl_aria256ofb_functions),
-    ALG(PROV_NAMES_ARIA_192_OFB, ossl_aria192ofb_functions),
-    ALG(PROV_NAMES_ARIA_128_OFB, ossl_aria128ofb_functions),
-    ALG(PROV_NAMES_ARIA_256_CFB, ossl_aria256cfb_functions),
-    ALG(PROV_NAMES_ARIA_192_CFB, ossl_aria192cfb_functions),
-    ALG(PROV_NAMES_ARIA_128_CFB, ossl_aria128cfb_functions),
-    ALG(PROV_NAMES_ARIA_256_CFB1, ossl_aria256cfb1_functions),
-    ALG(PROV_NAMES_ARIA_192_CFB1, ossl_aria192cfb1_functions),
-    ALG(PROV_NAMES_ARIA_128_CFB1, ossl_aria128cfb1_functions),
-    ALG(PROV_NAMES_ARIA_256_CFB8, ossl_aria256cfb8_functions),
-    ALG(PROV_NAMES_ARIA_192_CFB8, ossl_aria192cfb8_functions),
-    ALG(PROV_NAMES_ARIA_128_CFB8, ossl_aria128cfb8_functions),
-    ALG(PROV_NAMES_ARIA_256_CTR, ossl_aria256ctr_functions),
-    ALG(PROV_NAMES_ARIA_192_CTR, ossl_aria192ctr_functions),
-    ALG(PROV_NAMES_ARIA_128_CTR, ossl_aria128ctr_functions),
-#endif /* OPENSSL_NO_ARIA */
-#ifndef OPENSSL_NO_CAMELLIA
-    ALG(PROV_NAMES_CAMELLIA_256_ECB, ossl_camellia256ecb_functions),
-    ALG(PROV_NAMES_CAMELLIA_192_ECB, ossl_camellia192ecb_functions),
-    ALG(PROV_NAMES_CAMELLIA_128_ECB, ossl_camellia128ecb_functions),
-    ALG(PROV_NAMES_CAMELLIA_256_CBC, ossl_camellia256cbc_functions),
-    ALG(PROV_NAMES_CAMELLIA_192_CBC, ossl_camellia192cbc_functions),
-    ALG(PROV_NAMES_CAMELLIA_128_CBC, ossl_camellia128cbc_functions),
-    ALG(PROV_NAMES_CAMELLIA_128_CBC_CTS, ossl_camellia128cbc_cts_functions),
-    ALG(PROV_NAMES_CAMELLIA_192_CBC_CTS, ossl_camellia192cbc_cts_functions),
-    ALG(PROV_NAMES_CAMELLIA_256_CBC_CTS, ossl_camellia256cbc_cts_functions),
-    ALG(PROV_NAMES_CAMELLIA_256_OFB, ossl_camellia256ofb_functions),
-    ALG(PROV_NAMES_CAMELLIA_192_OFB, ossl_camellia192ofb_functions),
-    ALG(PROV_NAMES_CAMELLIA_128_OFB, ossl_camellia128ofb_functions),
-    ALG(PROV_NAMES_CAMELLIA_256_CFB, ossl_camellia256cfb_functions),
-    ALG(PROV_NAMES_CAMELLIA_192_CFB, ossl_camellia192cfb_functions),
-    ALG(PROV_NAMES_CAMELLIA_128_CFB, ossl_camellia128cfb_functions),
-    ALG(PROV_NAMES_CAMELLIA_256_CFB1, ossl_camellia256cfb1_functions),
-    ALG(PROV_NAMES_CAMELLIA_192_CFB1, ossl_camellia192cfb1_functions),
-    ALG(PROV_NAMES_CAMELLIA_128_CFB1, ossl_camellia128cfb1_functions),
-    ALG(PROV_NAMES_CAMELLIA_256_CFB8, ossl_camellia256cfb8_functions),
-    ALG(PROV_NAMES_CAMELLIA_192_CFB8, ossl_camellia192cfb8_functions),
-    ALG(PROV_NAMES_CAMELLIA_128_CFB8, ossl_camellia128cfb8_functions),
-    ALG(PROV_NAMES_CAMELLIA_256_CTR, ossl_camellia256ctr_functions),
-    ALG(PROV_NAMES_CAMELLIA_192_CTR, ossl_camellia192ctr_functions),
-    ALG(PROV_NAMES_CAMELLIA_128_CTR, ossl_camellia128ctr_functions),
-#endif /* OPENSSL_NO_CAMELLIA */
-#ifndef OPENSSL_NO_DES
-    ALG(PROV_NAMES_DES_EDE3_ECB, ossl_tdes_ede3_ecb_functions),
-    ALG(PROV_NAMES_DES_EDE3_CBC, ossl_tdes_ede3_cbc_functions),
-    ALG(PROV_NAMES_DES_EDE3_OFB, ossl_tdes_ede3_ofb_functions),
-    ALG(PROV_NAMES_DES_EDE3_CFB, ossl_tdes_ede3_cfb_functions),
-    ALG(PROV_NAMES_DES_EDE3_CFB8, ossl_tdes_ede3_cfb8_functions),
-    ALG(PROV_NAMES_DES_EDE3_CFB1, ossl_tdes_ede3_cfb1_functions),
-    ALG(PROV_NAMES_DES3_WRAP, ossl_tdes_wrap_cbc_functions),
-    ALG(PROV_NAMES_DES_EDE_ECB, ossl_tdes_ede2_ecb_functions),
-    ALG(PROV_NAMES_DES_EDE_CBC, ossl_tdes_ede2_cbc_functions),
-    ALG(PROV_NAMES_DES_EDE_OFB, ossl_tdes_ede2_ofb_functions),
-    ALG(PROV_NAMES_DES_EDE_CFB, ossl_tdes_ede2_cfb_functions),
-#endif /* OPENSSL_NO_DES */
-#ifndef OPENSSL_NO_SM4
-    ALG(PROV_NAMES_SM4_GCM, ossl_sm4128gcm_functions),
-    ALG(PROV_NAMES_SM4_CCM, ossl_sm4128ccm_functions),
-    ALG(PROV_NAMES_SM4_ECB, ossl_sm4128ecb_functions),
-    ALG(PROV_NAMES_SM4_CBC, ossl_sm4128cbc_functions),
-    ALG(PROV_NAMES_SM4_CTR, ossl_sm4128ctr_functions),
-    ALG(PROV_NAMES_SM4_OFB, ossl_sm4128ofb128_functions),
-    ALG(PROV_NAMES_SM4_CFB, ossl_sm4128cfb128_functions),
-    ALG(PROV_NAMES_SM4_XTS, ossl_sm4128xts_functions),
-#endif /* OPENSSL_NO_SM4 */
 #ifndef OPENSSL_NO_CHACHA
     ALG(PROV_NAMES_ChaCha20, ossl_chacha20_functions),
 #ifndef OPENSSL_NO_POLY1305
@@ -408,9 +312,6 @@ static const OSSL_ALGORITHM deflt_kdfs[] = {
 };
 
 static const OSSL_ALGORITHM deflt_keyexch[] = {
-#ifndef OPENSSL_NO_DH
-    { PROV_NAMES_DH, "provider=default", ossl_dh_keyexch_functions },
-#endif
 #ifndef OPENSSL_NO_EC
     { PROV_NAMES_ECDH, "provider=default", ossl_ecdh_keyexch_functions },
 #ifndef OPENSSL_NO_ECX
@@ -440,36 +341,14 @@ static const OSSL_ALGORITHM deflt_rands[] = {
 };
 
 static const OSSL_ALGORITHM deflt_signature[] = {
-#ifndef OPENSSL_NO_DSA
-    { PROV_NAMES_DSA, "provider=default", ossl_dsa_signature_functions },
-    { PROV_NAMES_DSA_SHA1, "provider=default", ossl_dsa_sha1_signature_functions },
-    { PROV_NAMES_DSA_SHA224, "provider=default", ossl_dsa_sha224_signature_functions },
-    { PROV_NAMES_DSA_SHA256, "provider=default", ossl_dsa_sha256_signature_functions },
-    { PROV_NAMES_DSA_SHA384, "provider=default", ossl_dsa_sha384_signature_functions },
-    { PROV_NAMES_DSA_SHA512, "provider=default", ossl_dsa_sha512_signature_functions },
-    { PROV_NAMES_DSA_SHA3_224, "provider=default", ossl_dsa_sha3_224_signature_functions },
-    { PROV_NAMES_DSA_SHA3_256, "provider=default", ossl_dsa_sha3_256_signature_functions },
-    { PROV_NAMES_DSA_SHA3_384, "provider=default", ossl_dsa_sha3_384_signature_functions },
-    { PROV_NAMES_DSA_SHA3_512, "provider=default", ossl_dsa_sha3_512_signature_functions },
-#endif
     { PROV_NAMES_RSA, "provider=default", ossl_rsa_signature_functions },
-#if !defined(OPENSSL_NO_RMD160) && !defined(FIPS_MODULE)
-    { PROV_NAMES_RSA_RIPEMD160, "provider=default", ossl_rsa_ripemd160_signature_functions },
-#endif
-    { PROV_NAMES_RSA_SHA1, "provider=default", ossl_rsa_sha1_signature_functions },
-    { PROV_NAMES_RSA_SHA224, "provider=default", ossl_rsa_sha224_signature_functions },
     { PROV_NAMES_RSA_SHA256, "provider=default", ossl_rsa_sha256_signature_functions },
     { PROV_NAMES_RSA_SHA384, "provider=default", ossl_rsa_sha384_signature_functions },
     { PROV_NAMES_RSA_SHA512, "provider=default", ossl_rsa_sha512_signature_functions },
-    { PROV_NAMES_RSA_SHA512_224, "provider=default", ossl_rsa_sha512_224_signature_functions },
     { PROV_NAMES_RSA_SHA512_256, "provider=default", ossl_rsa_sha512_256_signature_functions },
-    { PROV_NAMES_RSA_SHA3_224, "provider=default", ossl_rsa_sha3_224_signature_functions },
     { PROV_NAMES_RSA_SHA3_256, "provider=default", ossl_rsa_sha3_256_signature_functions },
     { PROV_NAMES_RSA_SHA3_384, "provider=default", ossl_rsa_sha3_384_signature_functions },
     { PROV_NAMES_RSA_SHA3_512, "provider=default", ossl_rsa_sha3_512_signature_functions },
-#ifndef OPENSSL_NO_SM3
-    { PROV_NAMES_RSA_SM3, "provider=default", ossl_rsa_sm3_signature_functions },
-#endif
 #ifndef OPENSSL_NO_EC
 #ifndef OPENSSL_NO_ECX
     { PROV_NAMES_ED25519, "provider=default", ossl_ed25519_signature_functions },
@@ -479,18 +358,12 @@ static const OSSL_ALGORITHM deflt_signature[] = {
     { PROV_NAMES_ED448ph, "provider=default", ossl_ed448ph_signature_functions },
 #endif
     { PROV_NAMES_ECDSA, "provider=default", ossl_ecdsa_signature_functions },
-    { PROV_NAMES_ECDSA_SHA1, "provider=default", ossl_ecdsa_sha1_signature_functions },
-    { PROV_NAMES_ECDSA_SHA224, "provider=default", ossl_ecdsa_sha224_signature_functions },
     { PROV_NAMES_ECDSA_SHA256, "provider=default", ossl_ecdsa_sha256_signature_functions },
     { PROV_NAMES_ECDSA_SHA384, "provider=default", ossl_ecdsa_sha384_signature_functions },
     { PROV_NAMES_ECDSA_SHA512, "provider=default", ossl_ecdsa_sha512_signature_functions },
-    { PROV_NAMES_ECDSA_SHA3_224, "provider=default", ossl_ecdsa_sha3_224_signature_functions },
     { PROV_NAMES_ECDSA_SHA3_256, "provider=default", ossl_ecdsa_sha3_256_signature_functions },
     { PROV_NAMES_ECDSA_SHA3_384, "provider=default", ossl_ecdsa_sha3_384_signature_functions },
     { PROV_NAMES_ECDSA_SHA3_512, "provider=default", ossl_ecdsa_sha3_512_signature_functions },
-#ifndef OPENSSL_NO_SM2
-    { PROV_NAMES_SM2, "provider=default", ossl_sm2_signature_functions },
-#endif
 #endif
 #ifndef OPENSSL_NO_ML_DSA
     { PROV_NAMES_ML_DSA_44, "provider=default", ossl_ml_dsa_44_signature_functions },
@@ -543,14 +416,10 @@ static const OSSL_ALGORITHM deflt_signature[] = {
 
 static const OSSL_ALGORITHM deflt_asym_cipher[] = {
     { PROV_NAMES_RSA, "provider=default", ossl_rsa_asym_cipher_functions },
-#ifndef OPENSSL_NO_SM2
-    { PROV_NAMES_SM2, "provider=default", ossl_sm2_asym_cipher_functions },
-#endif
     { NULL, NULL, NULL }
 };
 
 static const OSSL_ALGORITHM deflt_asym_kem[] = {
-    { PROV_NAMES_RSA, "provider=default", ossl_rsa_asym_kem_functions },
 #ifndef OPENSSL_NO_EC
 #ifndef OPENSSL_NO_ECX
     { PROV_NAMES_X25519, "provider=default", ossl_ecx_asym_kem_functions },
@@ -564,30 +433,16 @@ static const OSSL_ALGORITHM deflt_asym_kem[] = {
     { PROV_NAMES_ML_KEM_1024, "provider=default", ossl_ml_kem_asym_kem_functions },
 #if !defined(OPENSSL_NO_ECX)
     { PROV_NAMES_X25519MLKEM768, "provider=default", ossl_mlx_kem_asym_kem_functions },
-    { PROV_NAMES_X448MLKEM1024, "provider=default", ossl_mlx_kem_asym_kem_functions },
 #endif
 #if !defined(OPENSSL_NO_EC)
     { PROV_NAMES_SecP256r1MLKEM768, "provider=default", ossl_mlx_kem_asym_kem_functions },
     { PROV_NAMES_SecP384r1MLKEM1024, "provider=default", ossl_mlx_kem_asym_kem_functions },
 #endif
-#if !defined(OPENSSL_NO_SM2)
-    { PROV_NAMES_curveSM2MLKEM768, "provider=default", ossl_mlx_kem_asym_kem_functions },
-#endif
 #endif
     { NULL, NULL, NULL }
 };
 
 static const OSSL_ALGORITHM deflt_keymgmt[] = {
-#ifndef OPENSSL_NO_DH
-    { PROV_NAMES_DH, "provider=default", ossl_dh_keymgmt_functions,
-        PROV_DESCS_DH },
-    { PROV_NAMES_DHX, "provider=default", ossl_dhx_keymgmt_functions,
-        PROV_DESCS_DHX },
-#endif
-#ifndef OPENSSL_NO_DSA
-    { PROV_NAMES_DSA, "provider=default", ossl_dsa_keymgmt_functions,
-        PROV_DESCS_DSA },
-#endif
     { PROV_NAMES_RSA, "provider=default", ossl_rsa_keymgmt_functions,
         PROV_DESCS_RSA },
     { PROV_NAMES_RSA_PSS, "provider=default", ossl_rsapss_keymgmt_functions,
@@ -636,12 +491,6 @@ static const OSSL_ALGORITHM deflt_keymgmt[] = {
     { PROV_NAMES_CMAC, "provider=default", ossl_cmac_legacy_keymgmt_functions,
         PROV_DESCS_CMAC_SIGN },
 #endif
-#ifndef OPENSSL_NO_SM2
-    { PROV_NAMES_SM2, "provider=default", ossl_sm2_keymgmt_functions,
-        PROV_DESCS_SM2 },
-    { PROV_NAMES_curveSM2, "provider=default", ossl_curve_sm2_keymgmt_functions,
-        PROV_DESCS_curveSM2 },
-#endif
 #ifndef OPENSSL_NO_LMS
     { PROV_NAMES_LMS, "provider=default", ossl_lms_keymgmt_functions,
         PROV_DESCS_LMS },
@@ -656,8 +505,6 @@ static const OSSL_ALGORITHM deflt_keymgmt[] = {
 #if !defined(OPENSSL_NO_ECX)
     { PROV_NAMES_X25519MLKEM768, "provider=default", ossl_mlx_x25519_kem_kmgmt_functions,
         PROV_DESCS_X25519MLKEM768 },
-    { PROV_NAMES_X448MLKEM1024, "provider=default", ossl_mlx_x448_kem_kmgmt_functions,
-        PROV_DESCS_X448MLKEM1024 },
 #endif
 #if !defined(OPENSSL_NO_EC)
     { PROV_NAMES_SecP256r1MLKEM768, "provider=default", ossl_mlx_p256_kem_kmgmt_functions,
@@ -665,10 +512,6 @@ static const OSSL_ALGORITHM deflt_keymgmt[] = {
     { PROV_NAMES_SecP384r1MLKEM1024, "provider=default", ossl_mlx_p384_kem_kmgmt_functions,
         PROV_DESCS_SecP384r1MLKEM1024 },
 #endif
-#if !defined(OPENSSL_NO_SM2)
-    { PROV_NAMES_curveSM2MLKEM768, "provider=default", ossl_mlx_curve_sm2_kem_kmgmt_functions,
-        PROV_DESCS_curveSM2MLKEM768 },
-#endif
 #endif
 #ifndef OPENSSL_NO_SLH_DSA
     { PROV_NAMES_SLH_DSA_SHA2_128S, "provider=default", ossl_slh_dsa_sha2_128s_keymgmt_functions,
diff --git a/ssl/ssl_ciph.c b/ssl/ssl_ciph.c
index 61460c0f85..0f0b03769c 100644
--- a/ssl/ssl_ciph.c
+++ b/ssl/ssl_ciph.c
@@ -320,6 +320,24 @@ int ssl_load_ciphers(SSL_CTX *ctx)
     ctx->disabled_mkey_mask = 0;
     ctx->disabled_auth_mask = 0;
 
+    /*
+     * Never negotiate these, whichever providers are loaded: HMAC-SHA1 cipher
+     * suites, NULL encryption, AES-CCM with truncated 8 byte tags, DES, 3DES,
+     * Camellia, ARIA and SM4 encryption, anonymous and DSA authentication,
+     * RSA key transport, finite-field DHE, SRP, the GOST cipher suites and
+     * the PSK suites lacking ephemeral ECDH. ECDHE-PSK and TLSv1.3 PSK are
+     * unaffected.
+     */
+    ctx->disabled_enc_mask |= SSL_eNULL | SSL_AES128CCM8 | SSL_AES256CCM8
+        | SSL_DES | SSL_3DES | SSL_CAMELLIA | SSL_ARIAGCM
+        | SSL_SM4GCM | SSL_SM4CCM;
+    ctx->disabled_mac_mask |= SSL_SHA1 | SSL_GOST89MAC | SSL_GOST89MAC12
+        | SSL_MAGMAOMAC | SSL_KUZNYECHIKOMAC;
+    ctx->disabled_auth_mask |= SSL_aNULL | SSL_aDSS | SSL_aGOST01 | SSL_aGOST12;
+    ctx->disabled_mkey_mask |= SSL_kRSA | SSL_kDHE | SSL_kSRP
+        | SSL_kPSK | SSL_kRSAPSK | SSL_kDHEPSK
+        | SSL_kGOST | SSL_kGOST18;
+
     /*
      * We ignore any errors from the fetches below. They are expected to fail
      * if these algorithms are not available.
diff --git a/ssl/t1_lib.c b/ssl/t1_lib.c
index 4bc2748e97..f532be8c97 100644
--- a/ssl/t1_lib.c
+++ b/ssl/t1_lib.c
@@ -342,6 +342,30 @@ static int add_provider_groups(const OSSL_PARAM params[], void *data)
             ginf->mindtls = DTLS1_VERSION;
     }
 
+    /*
+     * Finite-field DH, SM2 and GOST groups, and the RFC 8422 deprecated EC
+     * groups (all sect* curves, and secp* below 256 bits, plus secp256k1),
+     * are never used for TLS, regardless of which provider advertises them.
+     * This also covers older FIPS modules built with tls-deprecated-ec.
+     * Skip them without fetching the algorithm, but report success so the
+     * provider's other groups are still added.
+     */
+    if (strcmp(ginf->algorithm, "DH") == 0
+        || (ginf->group_id >= OSSL_TLS_GROUP_ID_sect163k1
+            && ginf->group_id <= OSSL_TLS_GROUP_ID_secp256k1)
+        || (ginf->group_id >= OSSL_TLS_GROUP_ID_ffdhe2048
+            && ginf->group_id <= OSSL_TLS_GROUP_ID_ffdhe8192)
+        || strstr(ginf->algorithm, "SM2") != NULL
+        || strstr(ginf->tlsname, "SM2") != NULL
+        || OPENSSL_strncasecmp(ginf->algorithm, "gost", 4) == 0
+        || ginf->group_id == OSSL_TLS_GROUP_ID_curveSM2
+        || ginf->group_id == OSSL_TLS_GROUP_ID_curveSM2MLKEM768
+        || (ginf->group_id >= OSSL_TLS_GROUP_ID_gc256A
+            && ginf->group_id <= OSSL_TLS_GROUP_ID_gc512C)) {
+        ret = 1;
+        goto err;
+    }
+
     /*
      * Now check that the algorithm is actually usable for our property query
      * string. Regardless of the result we still return success because we have
@@ -599,6 +623,37 @@ static int add_provider_sigalgs(const OSSL_PARAM params[], void *data)
         goto err;
     }
 
+    /*
+     * DSA, SM2 and GOST signature schemes are never used for TLS, whichever
+     * provider advertises them. Report success so the provider's other
+     * signature schemes are still added.
+     */
+    keytype = inferred_keytype(sinf);
+    if (OPENSSL_strcasecmp(keytype, "DSA") == 0
+        || strstr(keytype, "SM2") != NULL
+        || OPENSSL_strncasecmp(keytype, "gost", 4) == 0
+        || OPENSSL_strncasecmp(sinf->sigalg_name, "dsa", 3) == 0
+        || OPENSSL_strncasecmp(sinf->sigalg_name, "sm2", 3) == 0
+        || OPENSSL_strncasecmp(sinf->sigalg_name, "gost", 4) == 0
+        || (sinf->sig_name != NULL
+            && (OPENSSL_strcasecmp(sinf->sig_name, "DSA") == 0
+                || strstr(sinf->sig_name, "SM2") != NULL
+                || OPENSSL_strncasecmp(sinf->sig_name, "gost", 4) == 0))
+        || sinf->code_point == TLSEXT_SIGALG_dsa_sha1
+        || sinf->code_point == TLSEXT_SIGALG_dsa_sha224
+        || sinf->code_point == TLSEXT_SIGALG_dsa_sha256
+        || sinf->code_point == TLSEXT_SIGALG_dsa_sha384
+        || sinf->code_point == TLSEXT_SIGALG_dsa_sha512
+        || sinf->code_point == TLSEXT_SIGALG_sm2sig_sm3
+        || sinf->code_point == TLSEXT_SIGALG_gostr34102012_256_intrinsic
+        || sinf->code_point == TLSEXT_SIGALG_gostr34102012_512_intrinsic
+        || sinf->code_point == TLSEXT_SIGALG_gostr34102012_256_gostr34112012_256
+        || sinf->code_point == TLSEXT_SIGALG_gostr34102012_512_gostr34112012_512
+        || sinf->code_point == TLSEXT_SIGALG_gostr34102001_gostr3411) {
+        ret = 1;
+        goto err;
+    }
+
     /*
      * Now check that the algorithm is actually usable for our property query
      * string. Regardless of the result we still return success because we have
@@ -2077,27 +2132,6 @@ static const uint16_t tls12_sigalgs[] = {
     TLSEXT_SIGALG_rsa_pkcs1_sha256,
     TLSEXT_SIGALG_rsa_pkcs1_sha384,
     TLSEXT_SIGALG_rsa_pkcs1_sha512,
-
-    TLSEXT_SIGALG_ecdsa_sha224,
-    TLSEXT_SIGALG_ecdsa_sha1,
-
-    TLSEXT_SIGALG_rsa_pkcs1_sha224,
-    TLSEXT_SIGALG_rsa_pkcs1_sha1,
-
-    TLSEXT_SIGALG_dsa_sha224,
-    TLSEXT_SIGALG_dsa_sha1,
-
-    TLSEXT_SIGALG_dsa_sha256,
-    TLSEXT_SIGALG_dsa_sha384,
-    TLSEXT_SIGALG_dsa_sha512,
-
-#ifndef OPENSSL_NO_GOST
-    TLSEXT_SIGALG_gostr34102012_256_intrinsic,
-    TLSEXT_SIGALG_gostr34102012_512_intrinsic,
-    TLSEXT_SIGALG_gostr34102012_256_gostr34112012_256,
-    TLSEXT_SIGALG_gostr34102012_512_gostr34112012_512,
-    TLSEXT_SIGALG_gostr34102001_gostr3411,
-#endif
 };
 
 static const uint16_t suiteb_sigalgs[] = {
@@ -2309,6 +2343,27 @@ static const uint16_t tls_default_sigalg[] = {
     0, /* SSL_PKEY_ED448 */
 };
 
+/*
+ * Signature schemes never used for TLS, even if a provider makes the digest
+ * or key type available for other purposes: anything involving MD5, SHA1,
+ * DSA, SM2, SM3 or GOST.
+ */
+static int tls_sigalg_blocked(int hash, int sig)
+{
+    return hash == NID_md5
+        || hash == NID_md5_sha1
+        || hash == NID_sha1
+        || hash == NID_sm3
+        || hash == NID_id_GostR3411_94
+        || hash == NID_id_GostR3411_2012_256
+        || hash == NID_id_GostR3411_2012_512
+        || sig == EVP_PKEY_DSA
+        || sig == EVP_PKEY_SM2
+        || sig == NID_id_GostR3410_2001
+        || sig == NID_id_GostR3410_2012_256
+        || sig == NID_id_GostR3410_2012_512;
+}
+
 int ssl_setup_sigalgs(SSL_CTX *ctx)
 {
     size_t i, cache_idx, sigalgs_len, enabled;
@@ -2356,6 +2411,11 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
             continue;
         }
 
+        if (tls_sigalg_blocked(lu->hash, lu->sig)) {
+            cache[i].available = 0;
+            continue;
+        }
+
         if (!EVP_PKEY_set_type(tmpkey, lu->sig)) {
             cache[i].available = 0;
             continue;
@@ -2387,6 +2447,8 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
         cache[cache_idx].maxdtls = -1;
         /* Compatibility with TLS 1.3 is checked on load */
         cache[cache_idx].available = istls;
+        if (tls_sigalg_blocked(cache[cache_idx].hash, cache[cache_idx].sig))
+            cache[cache_idx].available = 0;
         cache[cache_idx].advertise = 0;
         cache_idx++;
     }
-- 
2.53.0

